src/internal/observable/innerFrom.ts

```diff
@@ -64,12 +63,34 @@ export function fromInteropObservable<T>(obj: any) {
  * `from` conditionals because we *know* they're dealing with an array.
  * @param array The array to emit values from
  */
-export function fromArrayLike<T>(array: ArrayLike<T>) {
+function fromArrayLike<T>(array: ArrayLike<T>) {
   return new Observable((subscriber: Subscriber<T>) => {
     subscribeToArray(array, subscriber);
   });
 }
 
+/**
+ * Subscribes to an ArrayLike with a subscriber
+ * @param array The array or array-like to subscribe to
+ * @param subscriber
+ */
+export function subscribeToArray<T>(array: ArrayLike<T>, subscriber: Subscriber<T>) {
+  // Loop over the array and emit each value. Note two things here:
+  // 1. We're making sure that the subscriber is not closed on each loop.
+  //    This is so we don't continue looping over a very large array after
+  //    something like a `take`, `takeWhile`, or other synchronous unsubscription
+  //    has already unsubscribed.
+  // 2. In this form, reentrant code can alter that array we're looping over.
+  //    This is a known issue, but considered an edge case. The alternative would
+  //    be to copy the array before executing the loop, but this has
+  //    performance implications.
+  const length = array.length;
+  for (let i = 0; i < length && !subscriber.closed; i++) {
+    subscriber.next(array[i]);
+  }
```

Comment:
[Here](https://github.com/ReactiveX/rxjs/blob/f93e5c86a37af3b553a950b73eb346ee2806eddb/src/internal/observable/innerFrom.ts#L93-L95) we exit the function as soon as `subscriber.closed` becomes `true`. Does it make sense to do something similar here?

Reply:
I think this is covered in the `for` loop. This part: `&& !subscriber.closed` should ensure that.

Reply:
It's just that in [example](https://github.com/ReactiveX/rxjs/blob/f93e5c86a37af3b553a950b73eb346ee2806eddb/src/internal/observable/innerFrom.ts#L93-L95), when `subscriber.closed` becomes `true`, then `subscriber.complete()` is not called, but in `subscribeToArray` it is called.

Reply:
If `subscriber.complete()` gets called, but subscription is already closed, it won't make any difference. Subscriber will not get `complete` notification multiple times, so this is OK I think.

Reply:
Yes, I know that `complete` won't be notified several times. I just wanted to ask and suggest doing something like this:

```suggestion
  for (let i = 0; i < length; i++) {
    if (subscriber.closed) {
      return;
    }
    subscriber.next(array[i]);
  }
```

Nevermind, it's just my curiosity.

Reply:
@jakovljevic-mladen
I think it still needs to be fixed because `onStoppedNotification` will be called [here](https://github.com/ReactiveX/rxjs/blob/6c9930253a165303935ac096eeda65778dd8a3bc/src/internal/Subscriber.ts#L88).

```ts
import { config, Observable } from 'rxjs';

config.onStoppedNotification = (notification) => {
  console.log('onStoppedNotification', notification);
};

new Observable<never>((subscriber) => {
  subscriber.complete();
  subscriber.complete(); // log: onStoppedNotification {kind: "C", value: undefined, error: undefined}
}).subscribe();
```

Reply:
```suggestion
  for (let i = 0; i < length; i++) {
    if (subscriber.closed) {
      return;
    }
    subscriber.next(array[i]);
  }
```

I double-checked and made sure that if you do not return inside the loop, then `onStoppedNotification` will be called.

```ts
config.onStoppedNotification = (notification) => {
  console.log('onStoppedNotification', notification);
};

EMPTY.pipe(startWith(1, 2, 3), take(2)).subscribe();
```

Reply:
It is _mildly_ more efficient, I suppose. A conditional followed by a return, rather than a conditional followed by a function call, a conditional and a return. It's not likely to make any performance difference though, I'd guess. Even at high-speed.

Reply:
We are talking about a false positive `onStoppedNotification` call.

Reply:
The difference is that `return` prevents [`subscriber.complete()`](https://github.com/ReactiveX/rxjs/blob/95f8d958eb500be6d66263b06ed4263ad6808c19/src/internal/observable/innerFrom.ts#L91).